### ISSUE TITLE
Question about improving lexer

### ISSUE THREAD

**Issue by mezoni** [fkleon/math-expressions#64]
I wrote a `Lexer` for your grammar.
I did it with `parser_builder`.
https://github.com/mezoni/parser_builder

Below are benchmarks and source code for lexer. It needs build.
Place this script file in the `tool` folder and execute it.
A lexer will be generated.

Where it will be generated is specified in the script.
In our case, `tool/expr_parser.dart`.

```dart
Future<void> main(List<String> args) async {
  final context = Context();
  context.optimizeForSize = false;
  await fastBuild(context, [_parse], 'tool/expr_parser.dart',
      header: __header, publish: {'parse': _parse});
}
```

This lexer is very easy to change (improve, expand possibilities).
You can easily add support for variables like `foo2`. Currently, this is not supported (it seems to me).
The lexer does not remove spaces, and throws a `FormatException` at the location of the error.
In your case, the errors will be primitive, because it is a lexer (lexical analysis), but not a parser (grammar).

**Benchmark tests**
`'e^(x+2) +x * 2^2.5 * log(10,100) + (-(3.0^2.0))+- (1) + 1^1^1 * ln(10) / sqrt(10) * nrt(2,10) - ceil(1.2) + foo - baz_`

```
Time passed: 0.000, Test 'math_expressions lexer': 2193.664 ms
Time passed: 2.196, Test 'parser_builder lexer': 38.308 ms
Time passed: 2.235, Test 'math_expressions lexer': 2308.093 ms
Time passed: 4.543, Test 'parser_builder lexer': 4.19 ms
Time passed: 4.547, Test 'math_expressions lexer': 2310.533 ms
Time passed: 6.858, Test 'parser_builder lexer': 4.146 ms
Time passed: 6.862, Test 'math_expressions lexer': 2321.319 ms
Time passed: 9.184, Test 'parser_builder lexer': 4.347 ms
Time passed: 9.188, Test 'math_expressions lexer': 2306.278 ms
Time passed: 11.494, Test 'parser_builder lexer': 4.142 ms
```

Source cod of lexer builder:

```dart
import 'package:math_expressions/math_expressions.dart';
import 'package:parser_builder/branch.dart';
import 'package:parser_builder/bytes.dart';
import 'package:parser_builder/character.dart';
import 'package:parser_builder/combinator.dart';
import 'package:parser_builder/fast_build.dart';
import 'package:parser_builder/multi.dart';
import 'package:parser_builder/parser_builder.dart';
import 'package:parser_builder/sequence.dart';
import 'package:parser_builder/transformers.dart';

Future<void> main(List<String> args) async {
  final context = Context();
  context.optimizeForSize = false;
  await fastBuild(context, [_parse], 'tool/expr_parser.dart',
      header: __header, publish: {'parse': _parse});
}

const __header = r'''
// ignore_for_file: unused_local_variable

import 'package:math_expressions/math_expressions.dart';
import 'package:source_span/source_span.dart';
import 'package:tuple/tuple.dart';

const _source =
    'e^(x+2) +x * 2^2.5 * log(10,100) + (-(3.0^2.0))+- (1) + 1^1^1 * ln(10) / sqrt(10) * nrt(2,10) - ceil(1.2) + foo - baz_';

void main() {
  final count = 1000;
  final names = ['math_expressions lexer', 'parser_builder lexer'];
  final tests = [_test1, _test2];
  final sw = Stopwatch();
  var repeat = 5;
  sw.start();
  while (repeat-- > 0) {
    for (var i = 0; i < tests.length; i++) {
      final name = names[i];
      final test = tests[i];
      final seconds = (sw.elapsedMilliseconds / 1000).toStringAsFixed(3);
      final title = 'Time passed: $seconds, Test \'$name\'';
      _measure(title, 1, () => test(count));
    }
  }
}

void _measure(String name, int count, Function() f) {
  final sw = Stopwatch();
  sw.start();
  for (var i = 0; i < count; i++) {
    f();
  }
  sw.stop();
  final time = sw.elapsedMicroseconds / 1000;
  print('$name: $time ms');
}

void _test1(int count) {
  for (var i = 0; i < count; i++) {
    final lexer = Lexer();
    final res = lexer.tokenize(_source);
  }
}

void _test2(int count) {
  for (var i = 0; i < count; i++) {
    final res = parse(_source);
  }
}

''';

const _eof = Named('_eof', Eof<String>());

const _isIdentEnd = CharClasses([_isIdentStart, CharClass('[_]')]);

const _isIdentStart = CharClass('[a-zA-Z]');

const _number = Named(
    '_number',
    Map1(
        Recognize<String>(
          Sequence([
            Digit1(),
            Opt(Sequence([Tag('.'), Digit1()])),
          ]),
        ),
        Expr<Token>(['x'], 'Token({{x}}, TokenType.VAL)')));

const _parse = Named('_parse', Delimited(_ws, _tokenize, _eof));

const _token = Named(
    '_token',
    SwitchTag<Token>({
      '+': Skip(1, Expr.value("Token('+', TokenType.PLUS)")),
      '-': Skip(1, Expr.value("Token('-', TokenType.MINUS)")),
      '*': Skip(1, Expr.value("Token('*', TokenType.TIMES)")),
      '/': Skip(1, Expr.value("Token('/', TokenType.DIV)")),
      '%': Skip(1, Expr.value("Token('%', TokenType.MOD)")),
      '^': Skip(1, Expr.value("Token('^', TokenType.POW)")),
      '!': Skip(1, Expr.value("Token('!', TokenType.FACTORIAL)")),
      'sqrt': Skip(4, Expr.value("Token('sqrt', TokenType.SQRT)")),
      'log': Skip(3, Expr.value("Token('log', TokenType.LOG)")),
      'cos': Skip(3, Expr.value("Token('cos', TokenType.COS)")),
      'sin': Skip(3, Expr.value("Token('sin', TokenType.SIN)")),
      'tan': Skip(3, Expr.value("Token('tan', TokenType.TAN)")),
      'arccos': Skip(6, Expr.value("Token('arccos', TokenType.ACOS)")),
      'arcsin': Skip(6, Expr.value("Token('arcsin', TokenType.ASIN)")),
      'arctan': Skip(6, Expr.value("Token('arctan', TokenType.ATAN)")),
      'abs': Skip(3, Expr.value("Token('abs', TokenType.ABS)")),
      'ceil': Skip(4, Expr.value("Token('ceil', TokenType.CEIL)")),
      'floor': Skip(5, Expr.value("Token('floor', TokenType.FLOOR)")),
      'sgn': Skip(3, Expr.value("Token('sgn', TokenType.SGN)")),
      'ln': Skip(2, Expr.value("Token('ln', TokenType.LN)")),
      'e': Skip(1, Expr.value("Token('e', TokenType.EFUNC)")),
      'e^': Skip(2, Expr.value("Token('e', TokenType.EFUNC)")),
      '(': Skip(1, Expr.value("Token('(', TokenType.LBRACE)")),
      ')': Skip(1, Expr.value("Token(')', TokenType.RBRACE)")),
      '{': Skip(1, Expr.value("Token('{', TokenType.LBRACE)")),
      '}': Skip(1, Expr.value("Token('}', TokenType.RBRACE)")),
      ',': Skip(1, Expr.value("Token(',', TokenType.SEPAR)")),
      null: Alt2(_val, _var),
    }));

const _tokenize = Named(
    '_tokenize',
    Map1(ManyTill(Terminated(_token, _ws), _eof),
        Expr<List<Token>>(['x'], '{{x}}.item1')));

const _val = Named('_val', _number);

const _var = Named(
    '_var',
    Map1(
        Recognize<String>(
            Sequence([Satisfy(_isIdentStart), SkipWhile(_isIdentEnd)])),
        Expr<Token>(['x'], 'Token({{x}}, TokenType.VAR)')));

const _ws = Named('_ws', SkipWhile(CharClass('#x9 | #xA | #xD | #x20')));

typedef Expr<T> = ExprTransformer<T>;

```

**Comment by mezoni** [comment]
I've made a few changes already. This version of the code is no longer the most relevant.

**Comment by fkleon** [comment]
Thanks @mezoni for your suggestion, this approach looks interesting. I'll have a closer look at it over the weekend.

I actually started an alternative parser implementation based on a grammar defined via [dart-petitparser](https://github.com/petitparser/dart-petitparser) a while ago that I've not finished yet - it's on the [feature-parser-rewrite](https://github.com/fkleon/math-expressions/commits/feature-parser-rewrite) branch. It's still missing support for variables and custom function.

**Comment by mezoni** [comment]
Thanks, I'm already trying to implement a full featured parser for you.

**Comment by mezoni** [comment]
Little by little things are moving forward.
Not all implemented yet (not all rules added)

```
-----------------
x- => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ x-
  │ ^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ x-
  │ ^
  ╵
line 1, column 3: Expected: 'expression'
  ╷
1 │ x-
  │   ^
  ╵
-----------------
e^1 => e({1.0})
-----------------
e^(1) => e({1.0})
-----------------
-1^2^3*5^6%7/8 => ERROR
UnimplementedError
-----------------
!x => ERROR
FormatException: 
line 1, column 1: Expected: 'expression'
  ╷
1 │ !x
  │ ^
  ╵
-----------------
1+x*3-sqrt(5^2.0) => ((1.0 + (x * 3.0)) - sqrt({(5.0^2.0)}))
-----------------
sqrt+sqrt+1-2 => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ sqrt+sqrt+1-2
  │ ^^^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ sqrt+sqrt+1-2
  │ ^
  ╵
line 1, column 5: Expected: 'function arguments'
  ╷
1 │ sqrt+sqrt+1-2
  │     ^
  ╵
-----------------
1x => ERROR
FormatException: 
line 1, column 2: Expected: 'EOF'
  ╷
1 │ 1x
  │  ^
  ╵
-----------------
1+sqrt => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ 1+sqrt
  │ ^^^^^^
  ╵
line 1, column 3: Malformed expression
  ╷
1 │ 1+sqrt
  │   ^^^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ 1+sqrt
  │ ^
  ╵
line 1, column 3: Expected: 'expression'
  ╷
1 │ 1+sqrt
  │   ^
  ╵
line 1, column 7: Expected: 'function arguments'
  ╷
1 │ 1+sqrt
  │       ^
  ╵
-----------------
1+sqrt(2,3) => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ 1+sqrt(2,3)
  │ ^^^^^^^^^^^
  ╵
line 1, column 3: Malformed expression
  ╷
1 │ 1+sqrt(2,3)
  │   ^^^^^^^^^
  ╵
line 1, column 7: Argument list must contain 1 argument
  ╷
1 │ 1+sqrt(2,3)
  │       ^^^^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ 1+sqrt(2,3)
  │ ^
  ╵
line 1, column 3: Expected: 'expression'
  ╷
1 │ 1+sqrt(2,3)
  │   ^
  ╵
-----------------
nrt(x,y) => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ nrt(x,y)
  │ ^^^^^^^^
  ╵
line 1, column 4: The first argument must be a constant (number)
  ╷
1 │ nrt(x,y)
  │    ^^^^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ nrt(x,y)
  │ ^
  ╵
-----------------
addTwo(3 => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ addTwo(3
  │ ^^^^^^^^
  ╵
line 1, column 7: Malformed list of arguments
  ╷
1 │ addTwo(3
  │       ^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ addTwo(3
  │ ^
  ╵
line 1, column 7: Expected: 'function arguments'
  ╷
1 │ addTwo(3
  │       ^
  ╵
line 1, column 9: Expected: ')'
  ╷
1 │ addTwo(3
  │         ^
  ╵
-----------------
addTwo(3, => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ addTwo(3,
  │ ^^^^^^^^
  ╵
line 1, column 7: Malformed list of arguments
  ╷
1 │ addTwo(3,
  │       ^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ addTwo(3,
  │ ^
  ╵
line 1, column 7: Expected: 'function arguments'
  ╷
1 │ addTwo(3,
  │       ^
  ╵
line 1, column 9: Expected: ')'
  ╷
1 │ addTwo(3,
  │         ^
  ╵
-----------------
addTwo(3( => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ addTwo(3(
  │ ^^^^^^^^
  ╵
line 1, column 7: Malformed list of arguments
  ╷
1 │ addTwo(3(
  │       ^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ addTwo(3(
  │ ^
  ╵
line 1, column 7: Expected: 'function arguments'
  ╷
1 │ addTwo(3(
  │       ^
  ╵
line 1, column 9: Expected: ')'
  ╷
1 │ addTwo(3(
  │         ^
  ╵
-----------------
x- => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ x-
  │ ^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ x-
  │ ^
  ╵
line 1, column 3: Expected: 'expression'
  ╷
1 │ x-
  │   ^
  ╵
-----------------
5.x => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ 5.x
  │ ^^^
  ╵
line 1, column 1: Malformed decimal number
  ╷
1 │ 5.x
  │ ^^^
  ╵
line 1, column 3: Unexpected: 'x'
  ╷
1 │ 5.x
  │   ^
  ╵
line 1, column 1: Expected: 'expression', 'decimal number'
  ╷
1 │ 5.x
  │ ^
  ╵

```

**Comment by mezoni** [comment]
A lot of improvements had to be made to the `parser builder` today in order to get good locating results.
I haven't published them yet.
Now `parser builder` is better than it was yesterday.
The main job is to generate "good" errors.
Writing a primitive grammar is not difficult.
It's hard to implement the "right" grammar . So to speak, to make a fine adjustment. Like our universe. Scientists claim that the universe is finely tuned.
Thanks to your project, I have greatly improved the `parser builder` today.
Thank you for your work.

**Comment by mezoni** [comment]
I ran the simplest tests (directly in the parser script itself).
Below are the results.
Could you provide a list of expressions that need to be tested to check the parser work?
Including those that will contain syntax errors or  context-sensitive grammar  errors.

```
e^1
e^(1)
-1*3
-1^2^3*5^6%7/8
x!
x!*2
-x!*2
1+x*3-sqrt(5^2.0)
π
π*2
sin(1)
sin(90°)
sin(85°+x°)
sqrt+sqrt+1-2
1x
1+sqrt
1+sqrt(2,3)
nrt(x,y)
nrt(1)
addTwo(3
addTwo(3,
addTwo(3(
x-
5.x
-----------------
e^1 => e({1.0})
vars: {x: 5.0}
2.718281828459045
-----------------
e^(1) => e({1.0})
vars: {x: 5.0}
2.718281828459045
-----------------
-1*3 => ((-1.0) * 3.0)
vars: {x: 5.0}
-3.0
-----------------
-1^2^3*5^6%7/8 => (((((((-1.0)^2.0)^3.0) * 5.0)^6.0) % 7.0) / 8.0)
vars: {x: 5.0}
0.125
-----------------
x! => x!
vars: {x: 5.0}
120.0
-----------------
x!*2 => (x! * 2.0)
vars: {x: 5.0}
240.0
-----------------
-x!*2 => ({(-x)}! * 2.0)
 => ERROR
Invalid argument (Factorial): Negative values not supported.: -5.0
-----------------
1+x*3-sqrt(5^2.0) => ((1.0 + (x * 3.0)) - sqrt({(5.0^2.0)}))
vars: {x: 5.0}
11.0
-----------------
π => 3.141592653589793
vars: {x: 5.0}
3.141592653589793
-----------------
π*2 => (3.141592653589793 * 2.0)
vars: {x: 5.0}
6.283185307179586
-----------------
sin(1) => sin({1.0})
vars: {x: 5.0}
0.8414709848078965
-----------------
sin(90°) => sin({1.5707963267948966})
vars: {x: 5.0}
1.0
-----------------
sin(85°+x°) => sin({(1.4835298641951802 + (x * 0.017453292519943295))})
vars: {x: 5.0}
1.0
-----------------
sqrt+sqrt+1-2 => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ sqrt+sqrt+1-2
  │ ^^^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ sqrt+sqrt+1-2
  │ ^
  ╵
line 1, column 5: Expected: 'function arguments'
  ╷
1 │ sqrt+sqrt+1-2
  │     ^
  ╵
-----------------
1x => ERROR
FormatException: 
line 1, column 2: Expected: 'EOF'
  ╷
1 │ 1x
  │  ^
  ╵
-----------------
1+sqrt => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ 1+sqrt
  │ ^^^^^^
  ╵
line 1, column 3: Malformed expression
  ╷
1 │ 1+sqrt
  │   ^^^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ 1+sqrt
  │ ^
  ╵
line 1, column 3: Expected: 'expression'
  ╷
1 │ 1+sqrt
  │   ^
  ╵
line 1, column 7: Expected: 'function arguments'
  ╷
1 │ 1+sqrt
  │       ^
  ╵
-----------------
1+sqrt(2,3) => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ 1+sqrt(2,3)
  │ ^^^^^^^^^^^
  ╵
line 1, column 3: Malformed expression
  ╷
1 │ 1+sqrt(2,3)
  │   ^^^^^^^^^
  ╵
line 1, column 7: Argument list must contain 1 argument
  ╷
1 │ 1+sqrt(2,3)
  │       ^^^^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ 1+sqrt(2,3)
  │ ^
  ╵
line 1, column 3: Expected: 'expression'
  ╷
1 │ 1+sqrt(2,3)
  │   ^
  ╵
-----------------
nrt(x,y) => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ nrt(x,y)
  │ ^^^^^^^^
  ╵
line 1, column 4: The first argument must be a constant (number)
  ╷
1 │ nrt(x,y)
  │    ^^^^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ nrt(x,y)
  │ ^
  ╵
-----------------
nrt(1) => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ nrt(1)
  │ ^^^^^^
  ╵
line 1, column 4: Argument list must contain 2 arguments
  ╷
1 │ nrt(1)
  │    ^^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ nrt(1)
  │ ^
  ╵
-----------------
addTwo(3 => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ addTwo(3
  │ ^^^^^^^^
  ╵
line 1, column 7: Malformed list of arguments
  ╷
1 │ addTwo(3
  │       ^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ addTwo(3
  │ ^
  ╵
line 1, column 7: Expected: 'function arguments'
  ╷
1 │ addTwo(3
  │       ^
  ╵
line 1, column 9: Expected: ')'
  ╷
1 │ addTwo(3
  │         ^
  ╵
-----------------
addTwo(3, => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ addTwo(3,
  │ ^^^^^^^^
  ╵
line 1, column 7: Malformed list of arguments
  ╷
1 │ addTwo(3,
  │       ^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ addTwo(3,
  │ ^
  ╵
line 1, column 7: Expected: 'function arguments'
  ╷
1 │ addTwo(3,
  │       ^
  ╵
line 1, column 9: Expected: ')'
  ╷
1 │ addTwo(3,
  │         ^
  ╵
-----------------
addTwo(3( => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ addTwo(3(
  │ ^^^^^^^^
  ╵
line 1, column 7: Malformed list of arguments
  ╷
1 │ addTwo(3(
  │       ^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ addTwo(3(
  │ ^
  ╵
line 1, column 7: Expected: 'function arguments'
  ╷
1 │ addTwo(3(
  │       ^
  ╵
line 1, column 9: Expected: ')'
  ╷
1 │ addTwo(3(
  │         ^
  ╵
-----------------
x- => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ x-
  │ ^^
  ╵
line 1, column 1: Expected: 'expression'
  ╷
1 │ x-
  │ ^
  ╵
line 1, column 3: Expected: 'expression'
  ╷
1 │ x-
  │   ^
  ╵
-----------------
5.x => ERROR
FormatException: 
line 1, column 1: Malformed expression
  ╷
1 │ 5.x
  │ ^^^
  ╵
line 1, column 1: Malformed decimal number
  ╷
1 │ 5.x
  │ ^^^
  ╵
line 1, column 3: Unexpected: 'x'
  ╷
1 │ 5.x
  │   ^
  ╵
line 1, column 1: Expected: 'expression', 'decimal number'
  ╷
1 │ 5.x
  │ ^
  ╵
```

**Comment by mezoni** [comment]
To understand the essence of the matter.
I do this (develop a parser for you) to test [parser builder](https://github.com/mezoni/parser_builder) in real conditions.
After that, I will provide you with its source code.
You can do whatever you want with it. But I will always be glad to help in solving issues related to the use of [parser builder](https://github.com/mezoni/parser_builder) (to the best of my ability).
You can modify it yourself or users (contributors) can do that.
The original parser code is very simple, improvements can be made in a few minutes.
I will be glad if you are interested in my modest work (contribution).

P.S.
You can offer new features to your users as an experimental feature.
Alternative option.

P.S. P.S
The parser builder has already been tested by internal tests, as well as real data (many tens of megabytes) in external performance tests.

[test/test.dart](https://github.com/mezoni/parser_builder/blob/master/test/test.dart)
[JSON parser](https://github.com/mezoni/parser_builder/blob/master/example/example.dart)
[JSON parser builderr](https://github.com/mezoni/parser_builder/blob/master/tool/build_example.dart)
[CVS parser](https://github.com/mezoni/csv_parser)

**Comment by fkleon** [comment]
Looks like you had quite a bit of fun, that's a lot of information! I'm glad to hear that you could use my project to improve your parser builder.

> Could you provide a list of expressions that need to be tested to check the parser work?

There's a pretty comprehensive unit test suite that has plenty of examples of valid and invalid expressions. I see that you've already tried a lot of test expressions which is great.

> What do you think of the Unicode character support?

Awesome!

I still haven't had the chance to properly look at your parser builder tool, and the code you provided here. I'd love to see if it's possible to integrate your code on a branch as an alternative parser implementation. Then it would be fairly easy to run all tests across both implementations, and to have a play around with it. Unfortunately I have a few pretty busy weeks ahead, so I'll just have to wait until I can spare some time for my hobby projects :)

**Comment by mezoni** [comment]
Hello, Frederik! 
I created a pull request.
It does not make any changes to your project.
It adds a new parser. I named it `Parser2`.
It also adds one independent test, a parser builder (essentially a parser definition), and one example (a complete copy of your example).
That is, it doesn't break anything.
This feature also adds new features to the expression grammar.
And, of course, it allows you to simply expand the possibilities by modifying the grammar.

These modifications (improvements) can also be made by contributors of your project.
As far as I understand, there are developers who are interested in improvements in the project, judging by the issue tracker.

I am also always ready to help.
Any developer who is familiar with PEG parsers and parser combinators can easily understand the definition of a parser.
